### PR TITLE
crystal: change all crystal programs to use buildCrystalPackage and update pkgs

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

See the last comments for more details.

~~The `icr` change is straight forward and only included here to showcase the change in how we build crystal packages.~~

~~The crystal builder will now use a Makefile if one is included by the project, `shards` if we have the required files in place and otherwise build the specified binaries (the old behaviour).~~

Couple of other things to note:
1. we add a few more things to `nativeBuildInputs` that are required by several crystal programs.
2. upstream recommends not building with `--no-debug` so rip that out
3. support copying in a lockfile if the program doesn't have one (`lucky-cli` I'm looking at you)

~~The way I've done the conditionals where we combine bash conditionals with nix conditionals is *very* ugly and I'm not proud of that at all but it works.~~

Cc @manveru as the person we have to thank for having the crystal build infrastructure in the first place.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by manveru — APPROVED** [review]
Looks good to me :)

**Comment by peterhoeg** [comment]
Thanks for the feedback. I just made the build and install targets configurable, but otherwise this version is the same. I have `nixpkgs-fmt` running in a save hook, so that's the reason for all the noise.

**Comment by peterhoeg** [comment]
@GrahamcOfBorg build crystal2nix mint scry gitAndTools.thicket

**Comment by manveru** [comment]
Shards now depend on `buildCrystalPackage` to build, since they have introduced a shard as a dependency... so now we get infinite recursion with this.

**Comment by peterhoeg** [comment]
> now we get infinite recursion with this.

I saw that last night as well. Upstream was discussing distributing shards with the dependency vendored to avoid having to depend on itself. Maybe we could just do that to avoid the problem until they figure out what they want. I'm open to ideas.

**Comment by manveru** [comment]
It'd be nice if we allowed an attrset instead of a path for the `shardsFile`. That way you could inject your own versions of shards more easily. Maybe also allowing JSON format for the shards file, so the generator can be simpler and faster...

**Comment by peterhoeg** [comment]
I'm just seeing your comment about the attrset now, so that has not been addressed.

I hijacked my own PR to clean up the various crystal bits we carry to have them all up to date and using `buildCrystalPackage`. Few things to note:

1. we now run tests unless disabled
2. instead of the fugly bash/nix combo that tries to find the right thing to build, we simply make it configurable similar to how it works for python using a `format` attr.
3. I have a vendored version of `shards` which I think we should stick under the nixos/nix-community org for the future.
4. some of them are pinned against older crystal versions. I did that directly with each program instead of `all-packages.nix` to avoid forgetting about it at the next update cycle. I don't use `mint`, so I haven't tried that. But it compiles and the tests pass.
5. I had initially undone your work to move `shards` to `buildCrystalPackage` but with the configurable `format` we avoid the cyclical dependency.

**Comment by peterhoeg** [comment]
If nobody has any objections (@manveru @KimBurgess @filalex77) I'll be merging this soonish.

**Comment by KimBurgess** [comment]
Changes to Ameba look good :+1:

I've only had time for a shallow read so am not familiar with the current limitations, but working to move away from using that vendored shards repo seems like a neater long term. Possibly worth spinning up a separate issue for tracking / resolving that.

File: pkgs/development/tools/build-managers/shards/default.nix [hunk 1]
```diff
@@ -1,22 +1,27 @@
...
 crystal.buildCrystalPackage rec {
   pname = "shards";
   version = "0.10.0";
 
   src = fetchFromGitHub {
-    owner  = "crystal-lang";
-    repo   = "shards";
-    rev    = "v${version}";
-    sha256 = "1bjy3hcdqq8769bx73f3pwn26rnkj23dngyfbw4iv32bw23x1d49";
+    owner = "peterhoeg";
```
Comment (Mic92) [inline pkgs/development/tools/build-managers/shards/default.nix:8]: Are you the new maintainer of `shards`?

Comment (Mic92) [reply pkgs/development/tools/build-managers/shards/default.nix:8]: Ah. This is the vendored version. Maybe [nix-community](https://github.com/nix-community) would be a better place for this, so we can add more contributors easier.

Comment (peterhoeg) [reply pkgs/development/tools/build-managers/shards/default.nix:8]: I actually think we can do without the vendored version. Let me take a look.

Comment (peterhoeg) [reply pkgs/development/tools/build-managers/shards/default.nix]: And we can. Plus we now get man pages as well.

**Comment by peterhoeg** [comment]
@GrahamcOfBorg build ameba crystal2nix icr mint scry shards gitAndTools.thicket

**Comment by peterhoeg** [comment]
@GrahamcOfBorg build ameba crystal2nix icr mint scry shards gitAndTools.thicket

**Comment by manveru** [comment]
@peterhoeg seems like it still fails on missing `installManPage`.
If all builds pass, I'm all for merging this.

**Comment by peterhoeg** [comment]
I was working against unstable but when rebasing against master, I got the error too. Ahhh, well. It's fixed now.

**Comment by peterhoeg** [comment]
Thanks for your feedback guys, I'm merging this.

File: pkgs/development/tools/icr/default.nix [hunk 2]
```diff
@@ -1,28 +1,30 @@
...
 
-  nativeBuildInputs = [ shards which ];
+  # tests are failing due to our sandbox
+  doCheck = false;
 
-  doCheck = true;
-  checkTarget = "test";
+  postFixup = ''
+    wrapProgram $out/bin/icr \
+      --prefix PATH : ${lib.makeBinPath [ crystal shards makeWrapper which ]}
```
Comment (Artturin) [inline pkgs/development/tools/icr/default.nix:26]: the makeWrapper here looks like a mistake so i'll remove it

**Pull request merged**